src/Nethermind/Nethermind.Optimism/CL/P2P/PayloadByNumberProtocol.cs

```diff
@@ -0,0 +1,217 @@
+                return null;
+            default:
+                if (logger.IsWarn) logger.Warn($"{nameof(PayloadByNumberProtocol)}: Got unknown error code({res}). Payload number: {request}");
+                return null;
+        }
+        var version = BinaryPrimitives.ReadUInt32LittleEndian((await downChannel.ReadAsync(4)).Data.ToArray());
+
+        if (version != Version.Ecotone)
+        {
+            if (logger.IsWarn) logger.Warn($"{nameof(PayloadByNumberProtocol)}: Unsupported version. Payload number: {request}");
+            return null;
+        }
+
+        var decompressor = await ReadPayloadData(downChannel);
+
+        if (decompressor is null)
+        {
+            return null;
+        }
+
+        var buffer = ArrayPool<byte>.Shared.Rent(MaxResponseSizeBytes);
+        try
+        {
+            var bytesRead = decompressor.Read(buffer, 0, buffer.Length);
+            int totalRead = bytesRead;
+            while (bytesRead > 0)
+            {
+                if (totalRead > MaxResponseSizeBytes)
+                {
+                    return null;
+                }
+
+                bytesRead = decompressor.Read(buffer, totalRead, buffer.Length - totalRead);
+                totalRead += bytesRead;
+            }
+
+            return payloadDecoder.DecodePayload(buffer[..totalRead]);
+        }
+        catch (Exception e)
+        {
+            if (logger.IsWarn) logger.Warn($"{nameof(PayloadByNumberProtocol)}: Exception during payload decoding {e}");
+            return null;
+        }
+        finally
+        {
+            ArrayPool<byte>.Shared.Return(buffer);
+        }
+    }
+
+    private async Task<Stream?> ReadPayloadData(IReader reader)
+    {
+        List<ReadOnlySequenceStream> streams = new();
+        long bytesRead = 0;
+        var readResult = await reader.ReadAsync(MaxResponseSizeBytes, ReadBlockingMode.WaitAny);
+        while (readResult.Result == IOResult.Ok)
+        {
+            bytesRead += readResult.Data.Length;
+            if (bytesRead > MaxResponseSizeBytes)
+            {
+                return null;
+            }
+            streams.Add(new(readResult.Data));
+            readResult = await reader.ReadAsync(MaxResponseSizeBytes, ReadBlockingMode.WaitAny);
+        }
+
+        if (readResult.Result != IOResult.Ended)
+        {
+            return null;
+        }
+
+        bytesRead += readResult.Data.Length;
+        if (bytesRead > MaxResponseSizeBytes)
+        {
+            return null;
+        }
+        streams.Add(new(readResult.Data));
+
+        return new Snappier.SnappyStream(new CombinedStream(streams.GetEnumerator()), CompressionMode.Decompress);
+    }
+
+    private class CombinedStream : Stream
+    {
+        private readonly IEnumerator<Stream> _streams;
+        private Stream? _currentStream;
+
+        public CombinedStream(IEnumerator<Stream> streams)
```

Comment:
Probably you can do the following: 
- read all sequences using channel.ReadAllAsync()
- combine them into one using ReadOnlySequenceExtensions
- pass to a single stream
- when stream is read call sequence.CopyTo(buffer.AsSpan(offset, count); sequence= sequence.Slice(count);

One less stream like type; a bit simpler code

Reply:
We need to limit number of bytes that we reed to 10Mb. So we can't use ReadAll

Reply:
You can possibly add a counter. Hm let me add API that will make doing such requests more convenient